src/frontend/src/optimizer/plan_node/batch_sort_agg.rs

```diff
@@ -0,0 +1,157 @@
+use crate::optimizer::plan_node::ToLocalBatch;
+use crate::optimizer::property::{Distribution, Order, RequiredDist};
+
+#[derive(Debug, Clone)]
+pub struct BatchSortAgg {
+    pub base: PlanBase,
+    logical: LogicalAgg,
+    required_order: Order,
+}
+
+impl BatchSortAgg {
+    pub fn new(logical: LogicalAgg) -> Self {
+        let required_order = Order {
+            field_order: logical
+                .input()
+                .order()
+                .field_order
+                .iter()
+                .filter(|field_ord| {
+                    logical
+                        .group_key()
+                        .iter()
+                        .any(|g_k| *g_k == field_ord.index)
+                })
+                .cloned()
+                .collect(),
+        };
+        Self::with_required_order(logical, required_order)
+    }
+
+    fn with_required_order(logical: LogicalAgg, required_order: Order) -> Self {
+        let ctx = logical.base.ctx.clone();
+        let input = logical.input();
+        let input_dist = input.distribution();
+        let dist = match input_dist {
+            Distribution::HashShard(_) => logical
+                .i2o_col_mapping()
+                .rewrite_provided_distribution(input_dist),
+            d => d.clone(),
+        };
+        let base =
+            PlanBase::new_batch(ctx, logical.schema().clone(), dist, logical.order().clone());
+
+        assert_eq!(required_order.field_order.len(), logical.group_key().len());
+
+        BatchSortAgg {
+            base,
+            logical,
+            required_order,
+        }
+    }
+
+    pub fn agg_calls(&self) -> &[PlanAggCall] {
+        self.logical.agg_calls()
+    }
+
+    pub fn group_key(&self) -> &[usize] {
+        self.logical.group_key()
+    }
+}
+
+impl fmt::Display for BatchSortAgg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.logical.fmt_with_name(f, "BatchSortAgg")
+    }
+}
+
+impl PlanTreeNodeUnary for BatchSortAgg {
+    fn input(&self) -> PlanRef {
+        self.logical.input()
+    }
+
+    fn clone_with_input(&self, input: PlanRef) -> Self {
+        Self::with_required_order(
+            self.logical.clone_with_input(input),
+            self.required_order.clone(),
+        )
+    }
+}
+impl_plan_tree_node_for_unary! { BatchSortAgg }
+
+impl ToDistributedBatch for BatchSortAgg {
+    fn to_distributed(&self) -> Result<PlanRef> {
+        self.to_distributed_with_required(
+            &self.required_order,
+            &RequiredDist::shard_by_key(self.input().schema().len(), self.group_key()),
+        )
+    }
+
+    fn to_distributed_with_required(
+        &self,
+        required_order: &Order,
+        required_dist: &RequiredDist,
+    ) -> Result<PlanRef> {
+        let new_input = self
+            .input()
+            .to_distributed_with_required(required_order, required_dist)?;
+        Ok(self.clone_with_input(new_input).into())
+    }
```

Comment:
this method's `required_order` and `required_dist` means the requirement of `BatchSortAgg`'s output. so we should not implement this function but maintain the default implementation and implement the `to_distributed`. call `self.input().to_distributed_with_required` with self.required_order instead of this function's parameter

Reply:
Ah yeah of course. I drop my own required_order when called this way...
Thanks for the hint!